src/controller/python/chip/yaml/runner.py

```diff
@@ -0,0 +1,386 @@
+          'cluster': Name of cluster write attribute action is targeting.
+          'context': Contains test-wide common objects such as DataModelLookup instance.
+        Raises:
+          ParsingError: Raised if there is a benign error, and there is currently no
+            action to perform for this write attribute.
+          UnexpectedParsingError: Raised if there is an unexpected parsing error.
+        '''
+        super().__init__(test_step.label)
+        self._attribute_name = stringcase.pascalcase(test_step.attribute)
+        self._cluster = cluster
+        self._endpoint = test_step.endpoint
+        self._node_id = test_step.node_id
+        self._request_object = None
+
+        attribute = context.data_model_lookup.get_attribute(
+            self._cluster, self._attribute_name)
+        if attribute is None:
+            raise ParsingError(
+                f'WriteAttribute failed to find cluster:{self._cluster} '
+                f'Attribute:{self._attribute_name}')
+
+        if not test_step.arguments:
+            raise UnexpectedParsingError(f'WriteAttribute action does have arguments {self.label}')
+
+        args = test_step.arguments['values']
+        if len(args) != 1:
+            raise UnexpectedParsingError(f'WriteAttribute is trying to write multiple values')
+        request_data_as_dict = args[0]
+        try:
+            # TODO this is an ugly hack
+            request_data = Converter.convert_to_data_model_type(
+                request_data_as_dict['value'], attribute.attribute_type.Type)
+        except ValueError:
+            raise ParsingError('Could not covert yaml type')
+
+        # Create a cluster object for the request from the provided YAML data.
+        self._request_object = attribute(request_data)
+
+    def run_action(self, dev_ctrl: ChipDeviceCtrl) -> _ActionResult:
+        try:
+            resp = asyncio.run(
+                dev_ctrl.WriteAttribute(self._node_id, [(self._endpoint, self._request_object)]))
+        except chip.interaction_model.InteractionModelError:
+            # TODO Should we be doing the same thing as InvokeAction on InteractionModelError?
+            raise
+        if len(resp) == 1 and isinstance(resp[0], AttributeStatus):
+            if resp[0].Status == chip.interaction_model.Status.Success:
+                return _ActionResult(status=_ActionStatus.SUCCESS, response=None)
+            else:
+                return _ActionResult(status=_ActionStatus.ERROR, response=resp[0].Status)
+
+        # We always expecte the response to be a list of length 1, for that reason we return error
+        # here.
+        return _ActionResult(status=_ActionStatus.ERROR, response=None)
+
+
+class ReplTestRunner:
+    '''Test runner to encode/decode values from YAML test Parser for executing the TestStep.
+
+    Uses ChipDeviceCtrl from chip-repl to execute parsed YAML TestSteps.
+    '''
+
+    def __init__(self, test_spec_definition, dev_ctrl):
```

Comment:
These should have type hints to make it easier to do code completions

Reply:
Applies broadly to all functions added

Reply:
Chatted with Jerry offline, he is okay with this being done in a follow up PR. Right now the limitation of type hinting is that some stuff in the `scripts/` directory are not properly structured to created distributable python packages. That is coming in a follow up PR https://github.com/project-chip/connectedhomeip/pull/24140 that needs reviews from stake holders that are currently OOO for the holidays, once that is in we can add in proper type hinting.